README.md
# Ipdata.co API client for PHP

An API client to communicate with Ipdata.co.

## Install

```
composer require ipdata/api-client
```

The client is built upon PSR standards. We use PSR-18 for HTTP client and PSR-17 
for RequestFactory. These needs to be passed to the client's constructor. 

Example packages for PSR-17 and PSR-18:

```
composer require nyholm/psr7 symfony/http-client
```

```php
use Ipdata\ApiClient\Ipdata;
use Symfony\Component\HttpClient\Psr18Client;
use Nyholm\Psr7\Factory\Psr17Factory;

$httpClient = new Psr18Client();
$psr17Factory = new Psr17Factory();
$ipdata = new Ipdata('my_api_key', $httpClient, $psr17Factory);
```

## How to use

To send a geocode request you simply need to provide the IP address you are interested in. 

```php
$data = $ipdata->lookup('69.78.70.144');
echo json_encode($data, JSON_PRETTY_PRINT);
```

The output will be the response from the API server with one additional `status` field. 

```json
{
    "ip": "69.78.70.144",
    "is_eu": false,
    "city": null,
    "region": null,
    "region_code": null,
    "country_name": "United States",
    "country_code": "US",
    "continent_name": "North America",
    "continent_code": "NA",
    "latitude": 37.751,
    "longitude": -97.822,
    "postal": null,
    "calling_code": "1",
    "flag": "https:\/\/ipdata.co\/flags\/us.png",
    "emoji_flag": "\ud83c\uddfa\ud83c\uddf8",
    "emoji_unicode": "U+1F1FA U+1F1F8",
    "asn": {
        "asn": "AS6167",
        "name": "Cellco Partnership DBA Verizon Wireless",
        "domain": "verizonwireless.com",
        "route": "69.78.0.0\/16",
        "type": "business"
    },
    "carrier": {
        "name": "Verizon",
        "mcc": "310",
        "mnc": "004"
    },
    "languages": [
        {
            "name": "English",
            "native": "English"
        }
    ],
    "currency": {
        "name": "US Dollar",
        "code": "USD",
        "symbol": "$",
        "native": "$",
        "plural": "US dollars"
    },
    "time_zone": {
        "name": "America\/Chicago",
        "abbr": "CST",
        "offset": "-0600",
        "is_dst": false,
        "current_time": "2020-01-25T06:14:37.081772-06:00"
    },
    "threat": {
        "is_tor": false,
        "is_proxy": false,
        "is_anonymous": false,
        "is_known_attacker": false,
        "is_known_abuser": false,
        "is_threat": false,
        "is_bogon": false
    },
    "count": "6",
    "status": 200
}
```

If you are not interested in all the fields in the response, you may query only
the fields you want. 

```php
$data = $ipdata->lookup('69.78.70.144', ['longitude', 'latitude', 'country_name']);
echo json_encode($data, JSON_PRETTY_PRINT);
```

```json
{
    "longitude": -97.822,
    "latitude": 37.751,
    "country_name": "United States",
    "status": 200
}
```

### Bulk request

If you want to look up multiple IPs at the same time you may use the `bulkLookup()` function.


```php
$data = $ipdata->buildLookup(['1.1.1.1', '69.78.70.144'], ['longitude', 'latitude', 'country_name']);
echo json_encode($data, JSON_PRETTY_PRINT);
```

```json
{
    "0": {
        "longitude": 143.2104,
        "latitude": -33.494,
        "country_name": "Australia"
    },
    "1": {
        "longitude": -97.822,
        "latitude": 37.751,
        "country_name": "United States"
    },
    "status": 200
}
```

## Available Fields

A list of all the fields returned by the API is maintained at [Response Fields](https://docs.ipdata.co/api-reference/response-fields)

## Errors

A list of possible errors is available at [Status Codes](https://docs.ipdata.co/api-reference/status-codes)


CHANGELOG.md
# Change Log

## 0.1.1

### Fixed

- PHP8 support

## 0.1.0

Init release

tests/bootstrap.php
<?php


include dirname(__DIR__).'/vendor/autoload.php';

tests/IpdataTest.php
<?php

declare(strict_types=1);

namespace Ipdata\ApiClient\Tests;

use Http\Mock\Client as MockClient;
use Ipdata\ApiClient\Ipdata;
use Nyholm\Psr7\Factory\Psr17Factory;
use Nyholm\Psr7\Response;
use PHPUnit\Framework\TestCase;
use Psr\Http\Client\ClientInterface;
use Psr\Http\Message\ResponseInterface;

class IpdataTest extends TestCase
{
    public function testNoFields()
    {
        $httpClient = new MockClient();
        $httpClient->addResponse($this->createResponse());
        $ipdata = $this->createIpdata($httpClient);
        $ipdata->lookup('69.78.70.144');

        $request = $httpClient->getLastRequest();
        $this->assertEquals('GET', $request->getMethod());
        $this->assertEquals('/69.78.70.144', $request->getUri()->getPath());
        $this->assertEquals('api-key=secret_key', $request->getUri()->getQuery());
    }

    public function testOneField()
    {
        $httpClient = new MockClient();
        $httpClient->addResponse($this->createResponse());
        $ipdata = $this->createIpdata($httpClient);
        $ipdata->lookup('69.78.70.144', ['continent_code']);

        $request = $httpClient->getLastRequest();
        $this->assertEquals('/69.78.70.144', $request->getUri()->getPath());
        $this->assertEquals('api-key=secret_key&fields=continent_code', $request->getUri()->getQuery());
    }

    public function testMultipleFields()
    {
        $httpClient = new MockClient();
        $httpClient->addResponse($this->createResponse());
        $ipdata = $this->createIpdata($httpClient);
        $ipdata->lookup('69.78.70.144', ['country_name', 'threat']);

        $request = $httpClient->getLastRequest();
        $this->assertEquals('/69.78.70.144', $request->getUri()->getPath());
        $this->assertEquals('api-key=secret_key&fields=country_name,threat', urldecode($request->getUri()->getQuery()));
    }

    public function testBulkNoFields()
    {
        $httpClient = new MockClient();
        $httpClient->addResponse($this->createResponse());
        $ipdata = $this->createIpdata($httpClient);
        $ipdata->buildLookup(['8.8.8.8', '69.78.70.144']);

        $request = $httpClient->getLastRequest();
        $this->assertEquals('POST', $request->getMethod());
        $this->assertEquals('/bulk', $request->getUri()->getPath());
        $this->assertEquals('api-key=secret_key', $request->getUri()->getQuery());
        $this->assertEquals('["8.8.8.8","69.78.70.144"]', $request->getBody()->__toString());
    }

    public function testBulkOneField()
    {
        $httpClient = new MockClient();
        $httpClient->addResponse($this->createResponse());
        $ipdata = $this->createIpdata($httpClient);
        $ipdata->buildLookup(['8.8.8.8', '69.78.70.144'], ['continent_code']);

        $request = $httpClient->getLastRequest();
        $this->assertEquals('/bulk', $request->getUri()->getPath());
        $this->assertEquals('api-key=secret_key&fields=continent_code', $request->getUri()->getQuery());
        $this->assertEquals('["8.8.8.8","69.78.70.144"]', $request->getBody()->__toString());
    }

    public function testBulkMultipleFields()
    {
        $httpClient = new MockClient();
        $httpClient->addResponse($this->createResponse());
        $ipdata = $this->createIpdata($httpClient);
        $ipdata->buildLookup(['8.8.8.8', '69.78.70.144'], ['country_name', 'threat']);

        $request = $httpClient->getLastRequest();
        $this->assertEquals('/bulk', $request->getUri()->getPath());
        $this->assertEquals('api-key=secret_key&fields=country_name,threat', urldecode($request->getUri()->getQuery()));
        $this->assertEquals('["8.8.8.8","69.78.70.144"]', $request->getBody()->__toString());
    }

    /**
     * @dataProvider responseProvider
     */
    public function testStatusInResponse(int $statusCode, array $body)
    {
        $httpClient = new MockClient();
        $httpClient->addResponse($this->createResponse($body, $statusCode));
        $ipdata = $this->createIpdata($httpClient);

        $result = $ipdata->lookup('69.78.70.144');
        $this->assertArrayHasKey('status', $result);
        $this->assertEquals($statusCode, $result['status']);
    }

    public function responseProvider()
    {
        yield '200 response' => [200, ['foo' => 'bar']];
        yield '400 response' => [400, ['message' => '127.0.0.1 is a private IP address']];
        yield '401 response' => [401, ['message' => 'You have not provided a valid API Key.']];
        yield '403 response' => [403, ['message' => 'You have either exceeded your quota or that API key does not exist...']];
    }

    public function testParseNoneJsonResponse()
    {
        $httpClient = new MockClient();
        $response = new Response(200, ['Content-Type' => 'text/html'], 'Foobar');
        $httpClient->addResponse($response);
        $ipdata = $this->createIpdata($httpClient);

        $this->expectException(\RuntimeException::class);
        $ipdata->lookup('69.78.70.144');
    }

    public function testParseInvalidJsonResponse()
    {
        $httpClient = new MockClient();
        $response = new Response(200, ['Content-Type' => 'application/json'], 'Foobar');
        $httpClient->addResponse($response);
        $ipdata = $this->createIpdata($httpClient);

        $this->expectException(\RuntimeException::class);
        $ipdata->lookup('69.78.70.144');
    }

    public function testConstructWithDiscovery()
    {
        $ipdata = new Ipdata('secret_key');
        $this->assertInstanceOf(Ipdata::class, $ipdata);
    }

    private function createIpdata(ClientInterface $httpClient): Ipdata
    {
        return new Ipdata('secret_key', $httpClient, new Psr17Factory());
    }

    private function createResponse(array $data = [], int $statusCode = 200): ResponseInterface
    {
        if (empty($data)) {
            $data = [
                'ip' => '69.78.70.144',
                'is_eu' => false,
                'city' => null,
                'region' => null,
                'region_code' => null,
                'country_name' => 'United States',
                'country_code' => 'US',
                'continent_name' => 'North America',
                'continent_code' => 'NA',
                'latitude' => 37.751,
                'longitude' => -97.822,
                'postal' => null,
                'calling_code' => '1',
                'flag' => 'https://ipdata.co/flags/us.png',
                'emoji_flag' => "\ud83c\uddfa\ud83c\uddf8",
                'emoji_unicode' => 'U+1F1FA U+1F1F8',
                'asn' => [
                    'asn' => 'AS6167',
                    'name' => 'Cellco Partnership DBA Verizon Wireless',
                    'domain' => 'verizonwireless.com',
                    'route' => '69.78.0.0/16',
                    'type' => 'business',
                ],
                'carrier' => [
                    'name' => 'Verizon',
                    'mcc' => '310',
                    'mnc' => '004',
                ],
                'languages' => [
                    ['name' => 'English', 'native' => 'English'],
                ],
                'currency' => [
                    'name' => 'US Dollar',
                    'code' => 'USD',
                    'symbol' => '$',
                    'native' => '$',
                    'plural' => 'US dollars',
                ],
                'time_zone' => [
                    'name' => 'America/Chicago',
                    'abbr' => 'CST',
                    'offset' => '-0600',
                    'is_dst' => false,
                    'current_time' => '2020-01-25T06:11:31.771651-06:00',
                ],
                'threat' => [
                    'is_tor' => false,
                    'is_proxy' => false,
                    'is_anonymous' => false,
                    'is_known_attacker' => false,
                    'is_known_abuser' => false,
                    'is_threat' => false,
                    'is_bogon' => false,
                ],
                'count' => '3',
            ];
        }

        return new Response($statusCode, ['Content-Type' => 'application/json; charset=utf-8'], json_encode($data));
    }
}


src/Ipdata.php
<?php

declare(strict_types=1);

namespace Ipdata\ApiClient;

use Http\Discovery\Exception\NotFoundException;
use Http\Discovery\Psr17FactoryDiscovery;
use Http\Discovery\Psr18ClientDiscovery;
use Nyholm\Psr7\Factory\Psr17Factory;
use Psr\Http\Client\ClientInterface;
use Psr\Http\Message\RequestFactoryInterface;
use Psr\Http\Message\ResponseInterface;

/**
 * A small client to talk with Ipdata.co API.
 *
 * @author Tobias Nyholm <tobias.nyholm@gmail.com>
 */
class Ipdata
{
    private const BASE_URL = 'https://api.ipdata.co';

    /**
     * @var string|null
     */
    private $apiKey;

    /**
     * @var ClientInterface
     */
    private $httpClient;

    /**
     * @var RequestFactoryInterface
     */
    private $requestFactory;

    /**
     * Get an instance of the API client. Give it an API key, a PSR-18 client and a PSR-17 request factory.
     *
     * @param ClientInterface|null         $httpClient     if null, we will try to use php-http/discovery to find an installed client
     * @param RequestFactoryInterface|null $requestFactory if null, we will try to use php-http/discovery to find an installed factory
     */
    public function __construct(string $apiKey, ClientInterface $httpClient = null, RequestFactoryInterface $requestFactory = null)
    {
        if (null === $httpClient) {
            if (!class_exists(Psr18ClientDiscovery::class)) {
                throw new \LogicException(sprintf('You cannot use the "%s" without a PSR-18 HTTP client. Pass a "%s" as first argument to the constructor OR try running  "composer require php-http/discovery".', Ipdata::class, ClientInterface::class));
            }

            try {
                $httpClient = Psr18ClientDiscovery::find();
            } catch (NotFoundException $e) {
                throw new \LogicException('Could not find any installed HTTP clients. Try installing a package for this list: https://packagist.org/providers/psr/http-client-implementation', 0, $e);
            }
        }

        if (null === $requestFactory) {
            if (!class_exists(Psr17Factory::class) && !class_exists(Psr17FactoryDiscovery::class)) {
                throw new \LogicException(sprintf('You cannot use the "%s" as no PSR-17 request factory have been provided. Try running "composer require nyholm/psr7".', Ipdata::class));
            }

            try {
                $requestFactory = class_exists(Psr17Factory::class) ? new Psr17Factory() : Psr17FactoryDiscovery::findRequestFactory();
            } catch (NotFoundException $e) {
                throw new \LogicException('Could not find any PSR-17 factories. Try running "composer require nyholm/psr7".', 0, $e);
            }
        }

        $this->httpClient = $httpClient;
        $this->apiKey = $apiKey;
        $this->requestFactory = $requestFactory;
    }

    /**
     * @param array<string> $fields
     *
     * @throws \Psr\Http\Client\ClientExceptionInterface
     */
    public function lookup(string $ip, array $fields = []): array
    {
        $query = [
            'api-key' => $this->apiKey,
        ];

        if (!empty($fields)) {
            $query['fields'] = implode(',', $fields);
        }

        $request = $this->requestFactory->createRequest('GET', sprintf('%s/%s?%s', self::BASE_URL, $ip, http_build_query($query)));
        $response = $this->httpClient->sendRequest($request);

        return $this->parseResponse($response);
    }

    /**
     * Bulk lookup, requires paid subscription.
     *
     * @param array<string> $fields
     *
     * @throws \Psr\Http\Client\ClientExceptionInterface
     */
    public function buildLookup(array $ips, array $fields = []): array
    {
        $query = [
            'api-key' => $this->apiKey,
        ];

        if (!empty($fields)) {
            $query['fields'] = implode(',', $fields);
        }

        $request = $this->requestFactory->createRequest('POST', sprintf('%s/bulk?%s', self::BASE_URL, http_build_query($query)));
        $request->getBody()->write(json_encode($ips));
        $request = $request->withAddedHeader('Content-Type', 'text/plain');
        $response = $this->httpClient->sendRequest($request);

        return $this->parseResponse($response);
    }

    private function parseResponse(ResponseInterface $response): array
    {
        $body = $response->getBody()->__toString();
        if (0 !== strpos($response->getHeaderLine('Content-Type'), 'application/json')) {
            throw new \RuntimeException('Cannot convert response to array. Response has Content-Type:'.$response->getHeaderLine('Content-Type'));
        }

        $content = json_decode($body, true);
        if (JSON_ERROR_NONE !== json_last_error()) {
            throw new \RuntimeException(sprintf('Error (%d) when trying to json_decode response', json_last_error()));
        }

        $content['status'] = $response->getStatusCode();

        return $content;
    }
}
